File: internal/proxytag/matcher.go
package proxytag

import (
	"encoding/json"
	"os/exec"
	"errors"
)

type Matcher struct {
	Prefix string
	Suffix string
	Systemmate string
}

type Match struct {
	Prefix string
	Suffix string
	Body string
	Systemmate string
}

type OldMatch struct {
	// Name is the name of the systemmate, if the proxy method supplies this
	Name string `json:"name,omitempty"`
	// IntialSigil and EndSigil are the beginning and end non-alphanumeric
	// text to signify the speaker
	InitialSigil string `json:"initial_sigil,omitempty"`
	EndSigil     string `json:"end_sigil,omitempty"`
	// Method is the proxy method the scraper is looking for.
	Method string `json:"method"`
	// Body is the rest of what the systemmate said.
	Body string `json:"body"`
}

func (m OldMatch) Matchers() []Matcher {
	print("METHOD: " + m.Method)
	matchers := make([]Matcher, 0)
	switch m.Method {
	case "Nameslash":
		for _, sep := range []string{"\\", ":", "/", ">"} {
			matchers = append(matchers, Matcher{
				Prefix: m.Name + sep,
				Suffix: "",
				Systemmate: m.Name,
			})
		}
	case "Sigils":
		matchers = append(matchers, Matcher{
			Prefix: m.InitialSigil,
			Suffix: m.EndSigil,
			Systemmate: m.Name,
		})
	case "HalfSigilStart":
		matchers = append(matchers, Matcher{
			Prefix: m.InitialSigil,
			Suffix: "",
			Systemmate: m.Name,
		})
	}
	return matchers
}

func (m Matcher) String() string {
	return m.Prefix + " text " + m.Suffix
}

func MatchMessage(msg string, matchers []Matcher) (Match, error) {
	requestMatchers := make([]map[string]interface{}, len(matchers))
	for i, matcher := range matchers {
		requestMatchers[i] = map[string]interface{}{
			"matcherPrefix": matcher.Prefix,
			"matcherSuffix": matcher.Suffix,
			"matcherSystemMate": matcher.Systemmate,
		}
	}

	request := map[string]interface{}{
		"tag": "RequestMatchMessage",
		"contents": map[string]interface{}{
			"messageBody": msg,
			"messageMatchers": requestMatchers,
		},
	}

	response, err := matcherExec(request)
	if err != nil {
		return Match{}, err
	}

	switch tag := response["tag"].(string); tag {
	case "ResponseError":
		return Match{}, errors.New(response["contents"].(string))
	case "ResponseMatch":
		contents := response["contents"].(map[string]interface{})
		return Match{
			Prefix: contents["matchPrefix"].(string),
			Suffix: contents["matchSuffix"].(string),
			Body: contents["matchBody"].(string),
			Systemmate: contents["matchSystemMate"].(string),
		}, nil
	default:
		return Match{}, errors.New("error: unexpected response from proxy-matcher")
	}
}

func DetectMatcher(msg string, systemmate string) (Matcher, error) {
	request := map[string]interface{}{
		"tag": "RequestDetectMatcher",
		"contents": map[string]interface{}{
			"detectMatcherMessage": msg,
			"detectMatcherSystemMate": systemmate,
		},
	}

	response, err := matcherExec(request)
	if err != nil {
		return Matcher{}, err
	}

	switch tag := response["tag"].(string); tag {
	case "ResponseError":
		return Matcher{}, errors.New(response["contents"].(string))
	case "ResponseMatcher":
		contents := response["contents"].(map[string]interface{})
		suffix := ""
		if contents["matcherSuffix"] != nil {
			suffix = contents["matcherSuffix"].(string)
		}
		return Matcher{
			Prefix: contents["matcherPrefix"].(string),
			Suffix: suffix,
			Systemmate: contents["matcherSystemMate"].(string),
		}, nil
	default:
		return Matcher{}, errors.New("error: unexpected response from proxy-matcher")
	}
}

func matcherExec(request map[string]interface{}) (map[string]interface{}, error) {
	cmd := exec.Command("proxy-matcher")
	cmdStdin, err := cmd.StdinPipe()
	if err != nil {
		return nil, err
	}
	err = json.NewEncoder(cmdStdin).Encode(request)
	if err != nil {
		return nil, err
	}
	cmdStdin.Close()
	cmdStdout, err := cmd.Output()
	if err != nil {
		return nil, err
	}
	var response map[string]interface{}
	err = json.Unmarshal(cmdStdout, &response)
	if err != nil {
		return nil, err
	}
	return response, nil
}

File: internal/proxytag/Matcher.hs
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE OverloadedStrings #-}
module Main where

import Data.Text (Text)
import qualified Data.Text as Text
import qualified Data.Aeson as Aeson
import Data.Vector (Vector)
import GHC.Generics
import qualified Data.ByteString.Lazy as ByteString.Lazy
import Data.Monoid
import Data.Char

main :: IO ()
main = do
  input <- ByteString.Lazy.getContents
  let response =
        either (\x -> ResponseError (Text.append "error: " x)) id $ do
          req <-
            maybe
              (Left "could not decode message json")
              Right
              (Aeson.decode input)
          case req of
            RequestDetectMatcher detectReq -> do
              found <-
                maybe
                  (Left "no matcher detected")
                  Right
                  (detectMatcher detectReq)
              return (ResponseMatcher found)
            RequestMatchMessage msg -> do
              found <- maybe (Left "no match found") Right (messageMatch msg)
              return (ResponseMatch found)
      output = Aeson.encode response
  ByteString.Lazy.putStr output

data Request
  = RequestMatchMessage Message
  | RequestDetectMatcher DetectMatcher
  deriving (Eq, Read, Show, Generic)

instance Aeson.FromJSON Request

instance Aeson.ToJSON Request where
  toEncoding = Aeson.genericToEncoding Aeson.defaultOptions

data Response
  = ResponseError Text
  | ResponseMatch Match
  | ResponseMatcher Matcher
  deriving (Eq, Read, Show, Generic)

instance Aeson.ToJSON Response where
  toEncoding = Aeson.genericToEncoding Aeson.defaultOptions

instance Aeson.FromJSON Response

data DetectMatcher = DetectMatcher
  { detectMatcherMessage :: Text
  , detectMatcherSystemMate :: Text
  } deriving (Eq, Read, Show, Generic)

instance Aeson.ToJSON DetectMatcher where
  toEncoding = Aeson.genericToEncoding Aeson.defaultOptions

instance Aeson.FromJSON DetectMatcher

data Message = Message
  { messageBody :: Text
  , messageMatchers :: Vector Matcher
  } deriving (Eq, Read, Show, Generic)

instance Aeson.ToJSON Message where
  toEncoding = Aeson.genericToEncoding Aeson.defaultOptions

instance Aeson.FromJSON Message

data Matcher = Matcher
  { matcherPrefix :: Text
  , matcherSuffix :: Maybe Text
  , matcherSystemMate :: Text
  } deriving (Eq, Read, Show, Generic)

instance Aeson.ToJSON Matcher where
  toEncoding = Aeson.genericToEncoding Aeson.defaultOptions

instance Aeson.FromJSON Matcher

data Match = Match
  { matchPrefix :: Text
  , matchSuffix :: Maybe Text
  , matchBody :: Text
  , matchSystemMate :: Text
  } deriving (Eq, Read, Show, Generic)

instance Aeson.ToJSON Match where
  toEncoding = Aeson.genericToEncoding Aeson.defaultOptions

instance Aeson.FromJSON Match


matchRaw :: Matcher -> Text -> Maybe Match
matchRaw matcher body = do
  withoutPrefix <- Text.stripPrefix (matcherPrefix matcher) body
  let withoutSuffix = do
        suffix <- matcherSuffix matcher
        Text.stripSuffix suffix withoutPrefix
  case withoutSuffix of
    Nothing ->
      Just
        Match
          { matchPrefix = matcherPrefix matcher
          , matchSuffix = Nothing
          , matchBody = withoutPrefix
          , matchSystemMate = matcherSystemMate matcher
          }
    Just b ->
      Just
        Match
          { matchPrefix = matcherPrefix matcher
          , matchSuffix = matcherSuffix matcher
          , matchBody = b
          , matchSystemMate = matcherSystemMate matcher
          }

-- This is just matchRaw but it runs Text.strip on the input and output
match :: Matcher -> Text -> Maybe Match
match matcher body = do
  result <- matchRaw matcher (Text.strip body)
  Just (result {matchBody = Text.strip (matchBody result)})

messageMatch :: Message -> Maybe Match
messageMatch message =
  getAlt
    (foldMap
       (\matcher -> Alt (match matcher (messageBody message)))
       (messageMatchers message))

detectNamePrefixMatcher :: Text -> Maybe Matcher
detectNamePrefixMatcher msg =
  getAlt (foldMap (Alt . matcherForIndicator) validPrefixIndicators)
  where
    validPrefixIndicators = ["\\", "/", ":", ">"]
    matcherForIndicator indicator =
      case Text.splitOn indicator msg of
        [prefix, _] ->
          Just
            Matcher
              { matcherPrefix = Text.strip prefix <> indicator
              , matcherSuffix = Nothing
              , matcherSystemMate = ""
              }
        _ -> Nothing
  

detectSigilsMatcher :: Text -> Maybe Matcher
detectSigilsMatcher msg
  | Text.length msg == 1 && not (isAlphaNum (Text.head msg)) =
    Just
      Matcher
        {matcherPrefix = msg, matcherSuffix = Nothing, matcherSystemMate = ""}
  | Text.length msg == 2 && not (Text.any isAlphaNum msg) =
    Just
      Matcher
        { matcherPrefix = Text.singleton (Text.head msg)
        , matcherSuffix = Just (Text.singleton (Text.last msg))
        , matcherSystemMate = ""
        }
  | otherwise = Nothing
  

detectGenericMatcher :: Text -> Maybe Matcher
detectGenericMatcher msg =
  case Text.splitOn "text" msg of
    [prefix, suffix] ->
      Just
        Matcher
          { matcherPrefix = Text.strip prefix
          , matcherSuffix = Just (Text.strip suffix)
          , matcherSystemMate = ""
          }
    _ -> Nothing

detectMatcher :: DetectMatcher -> Maybe Matcher
detectMatcher detectReq = do
  foundMatcher <-
    getAlt
      (foldMap
         (\f -> Alt (f msg))
         [detectGenericMatcher, detectNamePrefixMatcher, detectSigilsMatcher])
  return foundMatcher {matcherSystemMate = sysMate}
  where
    msg = detectMatcherMessage detectReq
    sysMate = detectMatcherSystemMate detectReq
